specs/form-associated-custom-element.md

```diff
@@ -0,0 +1,114 @@
+
+#### Content attributes
+- `id: string`
+  - The id attribute of the component. Used for label association.
+- `value: string | File | FormData`
+  - When provided as a content attribute, value must be a string. When provided as an IDL property, value can be a `string`, `File`, or `FormData`
+- `autofocus: boolean`
+  - Boolean attribute. Will auto-focus the element on page load.
+- `disabled: boolean`
+  - Boolean attribute. Disables the form control, ensuring it doesn't participate in form submission
+- `name: string`
+  - The name of the element. Allows access by name from the associated form.
+- `required: boolean`
+  - Boolean attribute.
+
+### Risks and Challenges
+
+#### Retrieving labels
+Accessing the [`labels`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLInputElement/labels) property of a native input returns a `NodeList`. In cases where the form-association APIs are available, the implementation from `ElementInternals.labels` works as expected. In cases where we implement a proxy element though, we need to construct the label set from the following:
+  1. any parent `<label>`element. Can be retrieved from the proxy element's `labels` property
+  2. any `<label>` element with `[for="$id"]`, where $id is the `id` attribute *of the custom element*. Because the custom element does not reflect it's `id` attribute to the proxy element (that would result in two elements with the same ID in the DOM), the labels are not automatically associated to the `labels` property of the proxy.
+
+`NodeList`s are not constructable in JavaScript, and because we need to construct the label set in cases where we're using a proxy element, I'm purposing standardizing this property to `Node[]` , which AFAIK is as close to `NodeList` as we can get.
+
+#### Clicking labels
+Clicking a label of a native input element can have several side effects. First (and in general) it will focus the element. In cases like radio and checkbox, it changes the value of the input. There are likely other side-effects.
+
+How this works under the hood seems to be that clicking the label *also* invokes a click event on the element the label *labels*.
+
+However, because (in proxy element cases) the label's aren't *actually* associated to the element, the click event on the custom element isn't ever fired. This is the case for all label's using the `for` attribute to make the association. *Wrapping* labels (`<label><custom-element></label>`) *do* seem to fire the click event on the custom element.
+
+In order to *truly* support this scenario, we need to attach click event handlers to all labels for the elements. It *also* means that we need to attach new listeners when *new* labels get added to the DOM. This is the tricky part. We could use mutation observer, but my fear is that is heavy-handed for an edge-case scenario. 
```

Comment:
I would change first line to "In order to \*truly\* support this scenario, click event handlers must be attached to all labels for the elements."

Reply:
I see that the answer to my previous question...looks like we would need to eagerly evaluate the labels in order to hook up the events in this case.

What if we instead attached an event on the form that the element was within, caught it there, and then used that as a central place to handle labels? Then we don't need to worry about observing for new labels.

Reply:
That would probably catch *most* cases. It does remind me of a case I missed though (I'll add it) - the `form` content attribute can be assigned to the id of a form to be associated with, meaning the input doesn't *need* to be in a form. Probably a very edge-case but it is possible.

I guess we *could* say this component doesn't support that, in which case I think your suggestion would work.